### PR TITLE
docs(broker): Describe what Broker does and how to interact with it

### PR DESCRIPTION
**Pull request by lukehesluke**
Closes https://github.com/openactive/openactive-test-suite/issues/570

### CODE REVIEW

File: packages/openactive-broker-microservice/README.md [hunk 1]
```diff
@@ -294,3 +311,548 @@ Clients credentials can be defined as follows:
...
+
+* `type`: `orders` or `order-proposals`.
+* `bookingPartnerIdentifier`: Identifier of the Booking Partner whose [Orders Feed](https://openactive.io/open-booking-api/EditorsDraft/#dfn-orders-feed) or OrderProposals Feed to listen to. This is the key of the Booking Partner in the [`bookingPartners` configuration property](#bookingpartners) e.g. `primary`.
+* `uuid`: UUID of the Order or OrderProposal to listen for.
+
+**Response Type**: 🤖 JSON
+
+Create an [Order Listener](#two-phase-listeners) to listen for updates to the specified Order or Order Proposal.
+
+The response contains info that is useful to a human user for debugging, such as the `startingFeedPage`, which is the page of the feed that Broker Microservice will start listening from.
```
Comment (nickevansuk) [inline packages/openactive-broker-microservice/README.md:660]: Should a copy of an example response be included here, as with others above?

File: packages/openactive-broker-microservice/README.md [hunk 2]
```diff
@@ -294,3 +311,548 @@ Clients credentials can be defined as follows:
...
+This is used by some [Non-Implemented tests](../openactive-integration-tests/README.md#structure) to ensure that a test feature MUST be implemented if a certain criteria of Opportunity is found in the Booking System's data. e.g. the [`cancellation-window` feature](../openactive-integration-tests/test/features/cancellation/cancellation-window) has a non-implemented test that asserts that there are no Opportunities with cancellation windows defined.
+
+## Browser Automation for Auth Endpoints
+
+Broker Microservice also exposes endpoints created from the `setupBrowserAutomationRoutes(..)` function from [openactive-openid-test-client](../openactive-openid-test-client/).
+These endpoints are used by [Integration Tests](../openactive-integration-tests/) for tests which check the [OpenID Connect](https://openid.net/developers/how-connect-works/) authentication flow.
+
+## Concepts
+
+### Auth
```
Comment (nickevansuk) [inline packages/openactive-broker-microservice/README.md:773]:
```suggestion
### Authentication
```

File: packages/openactive-broker-microservice/README.md [hunk 3]
```diff
@@ -294,3 +311,548 @@ Clients credentials can be defined as follows:
...
+
+Broker Microservice sets up and maintains auth credentials for access to the [Booking System](#booking-system-under-test).
+
+[Integration Tests](../openactive-integration-tests/) can also obtain these credentials from [`GET /config`](#get-config) and use them to access the Booking System.
+
+These credentials are set up according to the auth strategy defined in the [`bookingPartners` configuration property](#bookingpartners).
```
Comment (nickevansuk) [inline packages/openactive-broker-microservice/README.md:779]:
```suggestion
These credentials are set up according to the authentication strategy defined in the [`bookingPartners` configuration property](#bookingpartners).
```

File: packages/openactive-broker-microservice/README.md [hunk 4]
```diff
@@ -294,3 +311,548 @@ Clients credentials can be defined as follows:
...
+
+### Opportunity Locks
+
+When running tests in [Random mode](../openactive-integration-tests/README.md#userandomopportunities), it is important that any Opportunity is only used in one test. This increases test coverage by ensuring that tests use a variety of different Opportunities and, crucially, it prevents unexpected behaviours when tests are run in parallel.
+
+Broker Microservice uses **Opportunity Locks** to ensure that an Opportunity is only used in one test. When an Opportunity is retrieved in Random mode, using [`POST /test-interface/datasets/:testDatasetIdentifier/opportunities`](#post-test-interfacedatasetstestdatasetidentifieropportunities), it is **locked**. Subsequent calls to get a random Opportunity with the same critera will not return the same Opportunity, as it is locked.
+
+In between runs of the tests in [Integration Tests](../openactive-integration-tests/), these locks need to be **released**, to ensure that all Opportunities are available for subsequent test runs.
```
Comment (nickevansuk) [inline packages/openactive-broker-microservice/README.md:826]:
```suggestion
In between runs of the tests in [Integration Tests](../openactive-integration-tests/), these locks are **released**, to ensure that all Opportunities are available for subsequent test runs.
```

Comment (nickevansuk) [reply packages/openactive-broker-microservice/README.md:826]: Just to clarify no additional action is required (as the release is automatic)

File: packages/openactive-broker-microservice/README.md [hunk 5]
```diff
@@ -294,3 +311,548 @@ Clients credentials can be defined as follows:
...
+* `harvesting`: Broker Microservice is harvesting or polling the [Booking System](#booking-system-under-test)'s RPDE feeds, which means that it is either in its [Initial Harvest](#initial-harvest) phase or it is polling the feeds to keep up to date.
+* `paused`: Broker Microservice is **not** harvesting or polling the [Booking System](#booking-system-under-test)'s RPDE feeds. This means that its data is not up to date with the Booking System's data, and so it is not available for use by [Integration Tests](../openactive-integration-tests/).
+
+### Opportunity ID Cache
+
+Broker Microservice keeps a cache of all the [Opportunities](https://openactive.io/open-booking-api/EditorsDraft/#dfn-opportunity) that it has harvested from the [Booking System](#booking-system-under-test)'s feeds.
+
+This cache is stored in various structures. For the purposes of interfacing with Broker Microservice, the most important one to know about is **Buckets**:
+
+#### Buckets
```
Comment (nickevansuk) [inline packages/openactive-broker-microservice/README.md:816]:
```suggestion
### Buckets
```

File: packages/openactive-broker-microservice/README.md [hunk 6]
```diff
@@ -294,3 +311,548 @@ Clients credentials can be defined as follows:
...
+
+A **Bucket** is a cache of Opportunity IDs that match a given [Opportunity Criteria](../test-interface-criteria/README.md). When Integration Tests sends a request to get a random Opportunity matching a given criteria, the Opportunity is fetched from these buckets.
```
Comment (nickevansuk) [inline packages/openactive-broker-microservice/README.md:818]:
```suggestion
A **Bucket** is a cache of Opportunity IDs that match a given [Opportunity Criteria](../test-interface-criteria/README.md). During harvesting, each Opportunity found in the feeds added to the relevant buckets.  When Integration Tests send a request to get a random Opportunity matching a given criteria, the Opportunity is fetched from these buckets.
```

File: packages/openactive-broker-microservice/src/util/opportunity-id-cache.js [hunk 7]
```diff
@@ -15,11 +15,15 @@ const { criteria } = require('@openactive/test-interface-criteria');
...
- *       -> Seller ID
- *         -> Set<Opportunity ID>
+ * -> {Criteria Name}
+ *   -> {Booking Flow}
+ *     -> {Opportunity Type}
+ *       -> contents
+ *         -> {Seller ID}
+ *           -> {Set<Opportunity ID>}
+ *       -> criteriaErrors
+ *         -> {constraint name (name of the constraint that failed)}
```
Comment (nickevansuk) [inline packages/openactive-broker-microservice/src/util/opportunity-id-cache.js:25]:
```suggestion
 *         -> {constraint name (name of the constraint that was not met)}
```

File: packages/openactive-integration-tests/README.md [hunk 8]
```diff
@@ -16,8 +16,9 @@ The results of this test suite when run against the reference implementation can
 To run `openactive-integration-tests` in separate terminal window to `openactive-broker-microservice`, from repository root:
 
 1. Ensure the [openactive-broker-microservice](../openactive-broker-microservice/) is running in another terminal window
-2. `export NODE_ENV=dev`
-3. `npm run start-tests`
+2. Specify Configuration file: `export NODE_ENV=dev`
+    * This is the command to use if using the `dev.json` config file, which is the default behaviour. See [Configuration](../../README.md#configuration) for more details)
```
Comment (nickevansuk) [inline packages/openactive-integration-tests/README.md:20]:
```suggestion
    * This is the command to use if using the `dev.json` config file. See [Configuration](../../README.md#configuration) for more details)
```

"default behaviour" could imply that this is the case without specifying `NODE_ENV=dev`, i.e. the "default" value if no other value is specified?

**Review by nickevansuk — APPROVED** [review]
Approved with some suggestions

File: packages/openactive-broker-microservice/README.md [hunk 9]
```diff
@@ -1,22 +1,31 @@
...
+1. [Harvests](#initial-harvest) data from the Booking System's Opportunity and Order RPDE feeds.
+    * This data is then [validated](#data-model-validation) and [cached](#opportunity-id-cache) and any [listeners](#two-phase-listeners) are notified.
+2. Fetch and parse data from the Booking System's [Dataset Site](https://openactive.io/dataset-api-discovery/EditorsDraft/).
+3. Set up and maintains [auth credentials](#auth) for access to the Booking System.
+
+It needs to be running, against a Booking System, in order for the [openactive-integration-tests](../openactive-integration-tests/) to run.
 
 ## Usage in separate terminal windows
 
 To run `openactive-broker-microservice` in separate terminal window to `openactive-integration-tests`, from repository root:
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md:14]: I've created a draft issue for this: https://github.com/orgs/openactive/projects/17/views/8?pane=issue&itemId=39029519 and so will not do it in this PR

File: packages/openactive-broker-microservice/README.md [hunk 10]
```diff
@@ -1,22 +1,31 @@
 ﻿# openactive-broker-microservice
 
-This Node.js microservice provides a service to harvest data and proxy calls to an Open Booking API
+Broker Microservice sits in front of a [Booking System](#booking-system-under-test), which is an implementation of the [Open Booking API specification](https://openactive.io/open-booking-api/EditorsDraft/), and provides the following services, to enable Test Suite:
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md]: Done

File: packages/openactive-broker-microservice/README.md [hunk 11]
```diff
@@ -1,22 +1,31 @@
...
+
+1. [Harvests](#initial-harvest) data from the Booking System's Opportunity and Order RPDE feeds.
+    * This data is then [validated](#data-model-validation) and [cached](#opportunity-id-cache) and any [listeners](#two-phase-listeners) are notified.
+2. Fetch and parse data from the Booking System's [Dataset Site](https://openactive.io/dataset-api-discovery/EditorsDraft/).
+3. Set up and maintains [auth credentials](#auth) for access to the Booking System.
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md]: Done

File: packages/openactive-broker-microservice/README.md [hunk 12]
```diff
@@ -294,3 +311,548 @@ Clients credentials can be defined as follows:
...
+          /* Unlike with TestOpporunityBookable, there were no ScheduledSessions
+          that matched this criteria, `TestOpportunityBookableNonFreeTaxNet`, so
+          instead of showing Opportunity numbers for each Seller, the status
+          endpoint shows a summary of why no Opportunities were found. */
+          "criteriaErrors": {
+            /* There are 892 ScheduledSessions which support Simple Booking Flow
+            and that do not satisfy the `TestOpportunityBookableNonFreeTaxNet`
+            criteria because they fail this contraint. That is to say that they
+            do not have seller tax mode TaxNet. */
+            "Seller Tax Mode Net": 892,
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md:434]: Done

File: packages/openactive-broker-microservice/README.md [hunk 13]
```diff
@@ -294,3 +311,548 @@ Clients credentials can be defined as follows:
...
+
+**Request params**:
+
+* `id`: The ID of the [Child Opportunity](#orphans).
+
+**Response Type**: 🤖 JSON
+
+Get an expanded* [Child Opportunity](#orphans) (e.g. a ScheduledSession or IndividualFacilityUseSlot) from Broker Microservice's cache, by its ID. Useful for debugging issues with the [Booking System](#booking-system-under-test)'s data.
+
+- *expanded: The Opportunity will have been expanded to include its parent Opportunity, if it has one.
```
[the remark this answers is not in the document]
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md]: Done

File: packages/openactive-broker-microservice/README.md (hunk 1 above)
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md]: Done

File: packages/openactive-broker-microservice/README.md (hunk 5 above)
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md:829]: This is supposed to be a sub-heading under Opportunity ID Cache

File: packages/openactive-broker-microservice/README.md (hunk 6 above)
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md:831]: Done

File: packages/openactive-integration-tests/README.md (hunk 8 above)
Comment (lukehesluke) [reply packages/openactive-integration-tests/README.md:20]: Good shout

File: packages/openactive-broker-microservice/README.md (hunk 2 above)
Comment (lukehesluke) [reply packages/openactive-broker-microservice/README.md:786]: My understanding is that Broker also manages authorization as well as authentication e.g. it maintains the booking partners' access tokens.

Comment (nickevansuk) [reply packages/openactive-broker-microservice/README.md:786]: Great point - something I've noticed is an issue in the booking spec too (and have raised https://github.com/openactive/open-booking-api/issues/240). Perhaps "Authentication and Authorisation" could be a good solution here too?

Comment (nickevansuk) [reply packages/openactive-broker-microservice/README.md:786]: However reflecting on this - in this specific case, is the configuration here more focussed on Authentication than Authorisation?

**Pull request merged**